### PR TITLE
Add line numbers and line highlighting to the RTF-formatter

### PR DESCRIPTION
**Pull request by jonascj**
Finally implemented line numbers and line highlighting as discussed in issue #1217.

Screenshot from 2024-02-25 23-01-27

If desirable, I could use f-strings instead of `%`-based string formatting while we're at it (maybe capitalize comments). I chose not to because it is not directly related to the issue.

Merging this would cause https://github.com/pygments/pygments/pull/2607 to need a minor rewrite. `\\sa0` would need to be added to [rtf.py#L269](https://github.com/jonascj/pygments/blob/83019842402a743b3b6828b73b61840c25bf0ae3/pygments/formatters/rtf.py#L269)

### CODE REVIEW

File: pygments/formatters/rtf.py [hunk 1]
```diff
@@ -62,6 +117,40 @@ def __init__(self, **options):
...
+
+        self.hl_color = options.get('hl_color', '')
+        if not self.hl_color:
+            self.hl_color = self.style.highlight_color
+
+        self.hl_lines = []
+        for lineno in get_list_opt(options, 'hl_lines', []):
+            try:
+                lineno = int(lineno)
+                if self.hl_linenostart :
```
Comment (Anteru) [inline pygments/formatters/rtf.py:136]: No extra space before `:` please

File: tests/test_rtf_formatter.py [hunk 2]
```diff
@@ -26,9 +31,9 @@ def _build_message(*args, **kwargs):
     result = _escape(kwargs.get('result', ''))
 
     if string is None:
-        string = ("The expected output of '{t}'\n"
-                  "\t\tShould be '{expected}'\n"
-                  "\t\tActually outputs '{result}'\n"
+        string = ("The expected output of '{t}'\n\n"
+                  "\t\tShould be '{expected}'\n\n"
+                  "\t\tActually outputs '{result}'\n\n"
```
Comment (Anteru) [inline tests/test_rtf_formatter.py:36]: Why do we have a double newline here now?

File: pygments/formatters/rtf.py [hunk 3]
```diff
@@ -42,6 +44,59 @@ class RtfFormatter(Formatter):
         default is 24 half-points, giving a size 12 font.
 
         .. versionadded:: 2.0
+
+    `linenos`
+        Turn on line numbering (default: ``False``).
+
+        .. versionadded:: 2.17
```
Comment (Anteru) [inline pygments/formatters/rtf.py:51]: 2.18 for all of those (sorry)

**Review by Anteru — CHANGES_REQUESTED** [review]

File: pygments/formatters/rtf.py (hunk 1 above)
Comment (jonascj) [reply pygments/formatters/rtf.py:136]: Of course not, good catch! I wonder why pylint didn't catch it ...

File: tests/test_rtf_formatter.py (hunk 2 above)
Comment (jonascj) [reply tests/test_rtf_formatter.py:36]: To improve readability of the pytest message when long lines wrap in the terminal / shell (see examples at the end of reply). 

Whether those custom messages are helpful or not is debatable. The `_build_message` function was made when all tests were single line input and single line output. When testing line numbering the input and output are multi-line, as are tests for the color table. 

I considered adding a command-line option to the tests which would dump RTF-test input, expected output and output to files in a specified directory: `tox -- tests/test_rtf_formatter.py --output-dir=/tmp/rtf-tests`. That would be more useful in my opinion, but requires modifications to `conftest.py` : https://docs.pytest.org/en/7.1.x/example/simple.html?highlight=addoption#pass-different-values-to-a-test-function-depending-on-command-line-options

**Double newline:**
```
E       AssertionError: The expected output of 's = "line1\nline2"\n'
E         
E                       Should be '{\rtf1\ansi\uc0\deff0{\fonttbl{\f0\fmodern\fprq1\fcharset0;}}\n{\colortbl;\n\red255\green0\blue0;\n\red0\green255\blu
e0;\n\red0\green0\blue255;\n}\n\f0\n\dntblnsbdb\n{\cf1 1  }{\cf3 s} = {\cf2 "}{\cf2 line1}{\cf2 \\n}{\cf2 line2}{\cf2 "}\par\n}\n'
E         
E                       Actually outputs '{\rtf1\ansi\uc0\deff0{\fonttbl{\f0\fmodern\fprq1\fcharset0;}}\n{\colortbl;\n\red255\green0\blue0;\n\red0\gree
n255\blue0;\n\red0\green0\blue255;\n}\n\f0\n\dntblnsbdb\n{\cf1 1  }{\cf3 s} = {\cf2 "}{\cf2 line1}{\cf2 \\n}{\cf2 line2}{\cf2 '
E         
E               (WARNING: Partial Output of Result!)
```

** Single newline:**
```
E       AssertionError: The expected output of 's = "line1\nline2"\n'
E                       Should be '{\rtf1\ansi\uc0\deff0{\fonttbl{\f0\fmodern\fprq1\fcharset0;}}\n{\colortbl;\n\red255\green0\blue0;\n\red0\green255\blu
e0;\n\red0\green0\blue255;\n}\n\f0\n\dntblnsbdb\n{\cf1 1  }{\cf3 s} = {\cf2 "}{\cf2 line1}{\cf2 \\n}{\cf2 line2}{\cf2 "}\par\n}\n'        
E                       Actually outputs '{\rtf1\ansi\uc0\deff0{\fonttbl{\f0\fmodern\fprq1\fcharset0;}}\n{\colortbl;\n\red255\green0\blue0;\n\red0\gree
n255\blue0;\n\red0\green0\blue255;\n}\n\f0\n\dntblnsbdb\n{\cf1 1  }{\cf3 s} = {\cf2 "}{\cf2 line1}{\cf2 \\n}{\cf2 line2}{\cf2 '
E               (WARNING: Partial Output of Result!)
```

File: pygments/formatters/rtf.py (hunk 3 above)
Comment (jonascj) [reply pygments/formatters/rtf.py:51]: No problem! `:%s/2.17/2.18/g`

File: tests/test_rtf_formatter.py (hunk 2 above)
Comment (Anteru) [reply tests/test_rtf_formatter.py:36]: Works for me, thanks!

**Comment by Anteru** [comment]
Merged, thanks a lot!

**Pull request merged**